### PR TITLE
unix, win: extend uv_udp_set_membership for IPv6 multicast.

### PR DESCRIPTION
**Pull request by snoj**
Added uv_udp_set_membership6.
Modified uv_udp_set_membership to call uv_udp_set_membership6
  when IPv6 is used.

Google groups discussion...not much there mainly easy access for myself. https://groups.google.com/forum/?fromgroups=#!topic/libuv/TxZmEnH2wP4

### CODE REVIEW

**Comment by snoj** [comment]
Ahhh shoot. Previous comments you guys put in are gone.

Anyway, made the changes that were recommended and refined the multicast binding. Now it doesn't send out on all interfaces if one is given. In order to do this though, I had to get the interface index which I did by extending uv_interface_address_s.

I'll start working on the Linux versions if there is nothing more.

**Comment by txdv** [comment]
If you come to the irc channel right now, I will aid you in this task.

File: src/win/udp.c [hunk 1]
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+  if (!(handle->flags & UV_HANDLE_BOUND) &&
+      uv_udp_bind6(handle, uv_addr_ip6_any_, 0) < 0) {
+    return -1;
+  }
+
+  memset(&mreq, 0, sizeof mreq);
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+  
+  if(interface_addr) {
```
Comment (bnoordhuis) [inline src/win/udp.c]: Style: space between keyword and paren.

File: src/win/udp.c [hunk 2]
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count).code != UV_OK) {
+      uv__set_sys_error(handle->loop, WSAGetLastError());
+      return -1;
```
Comment (bnoordhuis) [inline src/win/udp.c]: Fold into previous line.

File: src/win/udp.c [hunk 3]
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+    }
+
+    while(interfaces_count--) {
+      if(interfaces[interfaces_count].address.address4.sin_family == AF_INET6) {
+        if(memcmp(&interfaces[interfaces_count].address.address6.sin6_addr, &interface_addr_n, sizeof interface_addr_n) == 0) {
```
Comment (bnoordhuis) [inline src/win/udp.c]: Long line.

File: src/win/udp.c [hunk 4]
```diff
@@ -632,6 +631,80 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+    return -1;
+  }
+
+  memset(&mreq, 0, sizeof mreq);
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+  
+  if(interface_addr) {
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count).code != UV_OK) {
```
Comment (bnoordhuis) [inline src/win/udp.c]: Leaking interfaces?

File: src/win/util.c [hunk 5]
```diff
@@ -839,6 +839,9 @@ uv_err_t uv_interface_addresses(uv_interface_address_t** addresses_ptr,
         win_address->FirstUnicastAddress == NULL)
       continue;
 
+    /* Add four bytes for the interface index */
+    uv_address_buf_size += 4;
```
Comment (bnoordhuis) [inline src/win/util.c]: Use sizeof(fieldname) here.

Comment (bnoordhuis) [inline src/win/util.c]: By the way, is it necessary? It gets incremented with sizeof(uv_interface_address_t) further down below, right?

File: src/win/util.c [hunk 6]
```diff
@@ -901,6 +904,10 @@ uv_err_t uv_interface_addresses(uv_interface_address_t** addresses_ptr,
                                     (int) max_name_size,
                                     NULL,
                                     FALSE);
+
+    uv_address->ifIndex = 0;
+    uv_address->ifIndex = (win_address->IfIndex) ? win_address->IfIndex : win_address->Ipv6IfIndex;
```
Comment (bnoordhuis) [inline src/win/util.c]: Superfluous assignment, superfluous parens.

File: include/uv.h [hunk 7]
```diff
@@ -1433,6 +1450,7 @@ struct uv_cpu_info_s {
 struct uv_interface_address_s {
   char* name;
   int is_internal;
+  ULONG ifIndex;
```
Comment (bnoordhuis) [inline include/uv.h]: Use snake_case, not camelCase.

**Comment by snoj** [comment]
Super special thanks to @txdv for fixing the bad code style!

File: src/unix/internal.h [hunk 8]
```diff
@@ -244,3 +244,6 @@ static void uv__update_time(uv_loop_t* loop) {
 }
 
 #endif /* UV_UNIX_INTERNAL_H_ */
+
+/*  */
+#define UV_HANDLE_IPV6                          0x01000000
```
Comment (bnoordhuis) [inline src/unix/internal.h]: What's that doing there?

File: src/unix/udp.c [hunk 9]
```diff
@@ -524,6 +525,11 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
 
   int optname;
   struct ip_mreq mreq;
+
+  if (handle->flags & UV_HANDLE_IPV6) {
+    return uv_udp_set_membership6(handle, multicast_addr, interface_addr, membership);
```
Comment (bnoordhuis) [inline src/unix/udp.c]: If we're doing an implicit call then uv_udp_set_membership6 doesn't have to be public (and should be renamed to uv__udp_set_membership6 to indicate it's internal.)

File: src/unix/udp.c [hunk 10]
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
   return 0;
 }
 
+int uv_udp_set_membership6(uv_udp_t* handle, const char* multicast_addr,
+  const char* interface_addr, uv_membership membership) {
```
Comment (bnoordhuis) [inline src/unix/udp.c]: Arguments should line up, i.e.:

``` c
int uv_udp_set_membership6(uv_udp_t* handle,
                           const char* multicast_addr,
                           const char* interface_addr,
                           uv_membership membership
```

File: src/unix/udp.c (hunk 9 above)
Comment (bnoordhuis) [inline src/unix/udp.c]: EDIT: it should probably be static as well.

File: src/unix/udp.c [hunk 11]
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+
+  int optname;
+  struct ipv6_mreq mreq;
+  struct in6_addr multicast_addr_n;
+  struct in6_addr interface_addr_n;
+
```
Comment (bnoordhuis) [inline src/unix/udp.c]: Style: No newline.

File: src/unix/udp.c [hunk 12]
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+  memset(&mreq, 0, sizeof mreq);
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+
+  if (interface_addr) {
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count).code != UV_OK) {
+      uv__set_sys_error(handle->loop, UV_EINVAL);
+      return -1;
```
Comment (bnoordhuis) [inline src/unix/udp.c:520]: Style: fold into a single-line statement (`return uv__set_sys_error(handle->loop, UV_EINVAL);`) and drop the braces.

File: src/unix/udp.c [hunk 13]
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+    }
+
+    while (interfaces_count--) {
```
Comment (bnoordhuis) [inline src/unix/udp.c]: Micro-optimization and micro-style: use a `for (i = 0; i < interfaces_count; i++)` loop here.

Micro-optimization: CPU prefetchers are good at recognizing memory access patterns in ascending order but generally not when they're in descending order.

Micro-style: Prefer not mutating variables. It's a kind of future-proofing; if someone changes the code later on and assumes interfaces_count is factual, things will go badly wrong.

File: src/unix/udp.c [hunk 14]
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+      if (interfaces[interfaces_count].address.address4.sin_family == AF_INET6) {
+        if (memcmp(&interfaces[interfaces_count].address.address6.sin6_addr, &interface_addr_n, sizeof interface_addr_n) == 0) {
```
Comment (bnoordhuis) [inline src/unix/udp.c]: Style: Long line. Wrap at 80 characters max. I believe I pointed that out last night too. :-/

File: src/unix/udp.c [hunk 15]
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+  struct in6_addr interface_addr_n;
+
+  uv_interface_address_t* interfaces;
+  int interfaces_count;
+
+  memset(&mreq, 0, sizeof mreq);
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+
+  if (interface_addr) {
```
Comment (bnoordhuis) [inline src/unix/udp.c]: Style: interface_addr != NULL

File: src/unix/udp.c (hunk 14 above)
Comment (bnoordhuis) [inline src/unix/udp.c]: Also, using memcmp() here is potentially wrong. The struct may have holes in it that contain garbage.

File: src/unix/udp.c [hunk 16]
```diff
@@ -554,6 +560,69 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+    break;
+  case UV_LEAVE_GROUP:
+    optname = IPV6_DROP_MEMBERSHIP;
+    break;
+  default:
+    uv__set_sys_error(handle->loop, UV_EINVAL);
+    return -1;
+  }
+
+  if (setsockopt(handle->io_watcher.fd, IPPROTO_IPV6, optname, (void*) &mreq, sizeof mreq) == -1) {
```
Comment (bnoordhuis) [inline src/unix/udp.c]: Superfluous cast to void*.

File: test/test-udp-multicast-join6.c [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+static uv_udp_t server;
+static uv_udp_t client;
+
+static int cl_recv_cb_called;
+
+static int sv_send_cb_called;
+
+static int close_cb_called;
+
+static uv_buf_t alloc_cb(uv_handle_t* handle, size_t suggested_size) {
```
Comment (bnoordhuis) [inline test/test-udp-multicast-join6.c]: Style: Two newlines between declarations and functions.

File: src/unix/udp.c (hunk 16 above)
Comment (txdv) [inline src/unix/udp.c]: - space

File: src/unix/internal.h (hunk 8 above)
Comment (snoj) [inline src/unix/internal.h]: A copy of UV_HANDLE_IPV6 from src/win/internal.h. It was otherwise not available on my linux build.

Comment (bnoordhuis) [inline src/unix/internal.h]: Okay, but look closely _where_ you put it in src/unix/internal.h.

**Comment by snoj** [comment]
I believe the new forced commit takes care of most of the comments. I couldn't figure out a way compare ip addresses without more variables and conversions, perhaps I'll need to resort to that that.

**Comment by diversario** [comment]
I've tried to use multicast with IPv6 address but I keep getting `addMembership EINVAL`. Does this PR address that?

**Comment by snoj** [comment]
@diversario, When I started this, node/libuv did not support ipv6 multicast. I don't recall that EINVAL (see google group link at the top), but they may have changed the error codes since.

Since you're trying to use ipv6 multicasting, I think something like this would work for you. This code probably several version behind by now though. I or someone would need to review it.

**Comment by saghul** [comment]
Sorry this was left hanging. Can yoou please rebase it against current master?

File: src/unix/linux/linux-core.c [hunk 18]
```diff
@@ -717,6 +717,7 @@ uv_err_t uv_interface_addresses(uv_interface_address_t** addresses,
 #else
   struct ifaddrs *addrs, *ent;
   char ip[INET6_ADDRSTRLEN];
+
```
Comment (indutny) [inline src/unix/linux/linux-core.c]: Unnecessary whitespace change.

**Comment by snoj** [comment]
@saghul I'll attempt to rebase and run tests by sometime this coming weekend. But with it being christmas and with all the family get togethers, I can't guarantee anything.

**Comment by saghul** [comment]
No problem Josh!

**Comment by snoj** [comment]
Definitely not happening this weekend. To many changes have happened since writing this.

**Comment by saghul** [comment]
No need to hurry :-)

**Comment by mprokopowicz** [comment]
Any progress on this?  Can't wait for got this in node release.
It's critical for my project :)

**Comment by snoj** [comment]
Been working the last couple/few days with @mprokopowicz on this. Appears to be working, but when I rebased on the latest master, node broke. Tonight I'm hoping to rebase on the version node is using and going from there.

**Comment by snoj** [comment]
Done? Code compiles and ipv6 multicast tests completes without errors.

!ping @saghul ?

File: src/unix/udp.c [hunk 19]
```diff
@@ -35,6 +35,10 @@
 static void uv__udp_recvmsg(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
 static void uv__udp_sendmsg(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
 static int uv__udp_maybe_deferred_bind(uv_udp_t* handle, int domain);
+static int uv__udp_set_membership6(uv_udp_t* handle,
```
Comment (indutny) [inline src/unix/udp.c:38]: Whitespace at the end of the line

File: src/unix/udp.c [hunk 20]
```diff
@@ -35,6 +35,10 @@
 static void uv__udp_recvmsg(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
 static void uv__udp_sendmsg(uv_loop_t* loop, uv__io_t* w, unsigned int revents);
 static int uv__udp_maybe_deferred_bind(uv_udp_t* handle, int domain);
+static int uv__udp_set_membership6(uv_udp_t* handle, 
+                                   const char* multicast_addr,
+                                   const char* interface_addr,
```
Comment (indutny) [inline src/unix/udp.c:40]: Ditto

File: src/win/udp.c [hunk 21]
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+                            uv_membership membership) {
+  int optname;
+  int interfaces_count;
+  int i;
+  int err;
+  struct ipv6_mreq mreq;
+  struct in6_addr multicast_addr_n;
+  struct in6_addr interface_addr_n;
+  uv_interface_address_t* interfaces;
+
```
Comment (indutny) [inline src/win/udp.c:600]: Ditto

**Comment by indutny** [comment]
Got this on OSX:

```
../../src/unix/udp.c:549:15: error: use of undeclared identifier 'IPV6_ADD_MEMBERSHIP'
    optname = IPV6_ADD_MEMBERSHIP;
              ^
../../src/unix/udp.c:552:15: error: use of undeclared identifier 'IPV6_DROP_MEMBERSHIP'
    optname = IPV6_DROP_MEMBERSHIP;
              ^
```

**Comment by indutny** [comment]
Perhaps it is called `IPV6_JOIN_GROUP` and `IPV6_LEAVE_GROUP` on osx?

**Comment by indutny** [comment]
I'd suggest adding `UV__IPV6_JOIN_GROUP` and `UV__IPV6_LEAVE_GROUP` in `src/unix/internal.h` and setting them to the appropriate value

File: test/test-udp-multicast-join6.c [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+
+TEST_IMPL(udp_multicast_join6) {
+  int r;
+  uv_udp_send_t req;
+  uv_buf_t buf;
+  struct sockaddr_in6 addr;
+
+  ASSERT(0 == uv_ip6_addr("::1", TEST_PORT, &addr));
+
+  r = uv_udp_init(uv_default_loop(), &server);
```
Comment (indutny) [inline test/test-udp-multicast-join6.c:107]: Please do ASSERT(0 == ...) everywhere

File: src/unix/udp.c [hunk 23]
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+
+  memset(&mreq, 0, sizeof mreq);
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+
+  if (interface_addr != NULL) {
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count) != 0)
+      return -1;
+      /*return uv__set_sys_error(handle->loop, UV_EINVAL);*/
```
Comment (indutny) [inline src/unix/udp.c:521]: This is a bad idea :)

File: src/unix/udp.c [hunk 24]
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+  uv_interface_address_t* interfaces;
+
+  memset(&mreq, 0, sizeof mreq);
+  memset(&multicast_addr_n, 0, sizeof multicast_addr_n);
+  memset(&interface_addr_n, 0, sizeof interface_addr_n);
+
+  if (interface_addr != NULL) {
+    uv_inet_pton(AF_INET6, interface_addr, &interface_addr_n);
+    if (uv_interface_addresses(&interfaces, &interfaces_count) != 0)
+      return -1;
```
Comment (indutny) [inline src/unix/udp.c:520]: Just return `-errno`

File: src/unix/udp.c [hunk 25]
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+                   sizeof interface_addr_n) == 0) {
+          mreq.ipv6mr_interface = interfaces[i].if_index;
+          break;
+        }
+      }
+    }
+
+    if (mreq.ipv6mr_interface == 0) {
+      /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+      return -1;
```
Comment (indutny) [inline src/unix/udp.c:536]: Ditto

File: src/unix/udp.c [hunk 26]
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+    }
+  }
+
+  if (uv_inet_pton(AF_INET6, multicast_addr, &multicast_addr_n) != 0) {
+    /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+    return -1;
```
Comment (indutny) [inline src/unix/udp.c:542]: Ditto

File: src/unix/udp.c [hunk 27]
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+      }
+    }
+
+    if (mreq.ipv6mr_interface == 0) {
+      /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+      return -1;
+    }
+  }
+
+  if (uv_inet_pton(AF_INET6, multicast_addr, &multicast_addr_n) != 0) {
```
Comment (indutny) [inline src/unix/udp.c:540]: And no need in braces if there will be one line

File: src/unix/udp.c [hunk 28]
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+  switch (membership) {
+  case UV_JOIN_GROUP:
+    optname = IPV6_ADD_MEMBERSHIP;
+    break;
+  case UV_LEAVE_GROUP:
+    optname = IPV6_DROP_MEMBERSHIP;
+    break;
+  default:
+    /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+    return -1;
```
Comment (indutny) [inline src/unix/udp.c:556]: Ditto

Comment (indutny) [inline src/unix/udp.c:556]: Probably should be an assertion, or UV_ENOTSUP

File: src/unix/udp.c [hunk 29]
```diff
@@ -488,6 +497,73 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+  }
+
+  if (setsockopt(handle->io_watcher.fd, IPPROTO_IPV6, optname, &mreq, sizeof mreq) != 0) {
+    /*uv__set_artificial_error(handle->loop, UV_EINVAL);*/
+    return -1;
```
Comment (indutny) [inline src/unix/udp.c:561]: Dito

File: src/win/udp.c [hunk 30]
```diff
@@ -539,7 +539,7 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
   struct ip_mreq mreq;
 
   /* If the socket is unbound, bind to inaddr_any. */
-  if (!(handle->flags & UV_HANDLE_BOUND)) {
+  if (!(handle->flags & UV_HANDLE_IPV6) && !(handle->flags & UV_HANDLE_BOUND)) {
```
Comment (indutny) [inline src/win/udp.c:542]: `!(handle->flags & (UV_HANDLE_IPV6 | UV_HANDLE_BOUND))`

File: src/win/udp.c [hunk 31]
```diff
@@ -570,7 +570,7 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
       optname = IP_DROP_MEMBERSHIP;
       break;
     default:
-      return UV_EINVAL;
+      return UV__EINVAL;
```
Comment (indutny) [inline src/win/udp.c:573]: Why?

File: src/win/udp.c [hunk 32]
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+  int interfaces_count;
+  int i;
+  int err;
+  struct ipv6_mreq mreq;
+  struct in6_addr multicast_addr_n;
+  struct in6_addr interface_addr_n;
+  uv_interface_address_t* interfaces;
+    
+  if (!(handle->flags & UV_HANDLE_IPV6)) {
+	  return -1;
```
Comment (indutny) [inline src/win/udp.c:602]: Ditto

File: src/win/udp.c [hunk 33]
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+                   &interface_addr_n,
+                   sizeof interface_addr_n) == 0) {
+          mreq.ipv6mr_interface = interfaces[i].if_index;
+          break;
+        }
+      }
+    }
+
+    if (mreq.ipv6mr_interface == 0) {
+      return -1;
```
Comment (indutny) [inline src/win/udp.c:638]: Should return some meaningful error here

File: src/win/udp.c [hunk 34]
```diff
@@ -585,6 +585,87 @@ int uv_udp_set_membership(uv_udp_t* handle, const char* multicast_addr,
...
+
+  switch (membership) {
+    case UV_JOIN_GROUP:
+      optname = IPV6_ADD_MEMBERSHIP;
+      break;
+    case UV_LEAVE_GROUP:
+      optname = IPV6_DROP_MEMBERSHIP;
+      break;
+    default:
+      return -1;
```
Comment (indutny) [inline src/win/udp.c:655]: And there

**Comment by indutny** [comment]
As another note, I do get `EADDRNOTAVAIL` on OSX and test is failing.

**Comment by indutny** [comment]
Please fix all that things and I will finish reviewing

**Comment by indutny** [comment]
And generally it looks quite neat ;)

**Comment by saghul** [comment]
The UV_HANDLE_IPV6 flag is only used for UDP, maybe you could rename it to something like UV_UDP_IPV6?

**Comment by saghul** [comment]
I'd name the extra field you added to uv_interface_address_s just "index". Also, AFAIS it's only set on Linux and Windows. Don't other platforms need it as well?

**Comment by saghul** [comment]
Instead of branching out to a ipv6 version of the function and continuing on the same otherwise, I'd split uv_udp_set_membership into 2: uv__udp_set_membership4/6.

PS: sorry for the brevity, I'm on mobile.

File: src/unix/udp.c (hunk 27 above)
Comment (indutny) [inline src/unix/udp.c:540]: Also in all other places in diff.

**Comment by mprokopowicz** [comment]
Hi,

i was mailing with snoj and unfortunately he won't have time to finish this commit. Is any one here who can handle it?

**Comment by saghul** [comment]
I'll take over.

**Comment by mprokopowicz** [comment]
@saghul Awsome news :) Can U provide any estimates? (No pressure, just curious... )

**Comment by saghul** [comment]
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA1

On 03/14/2014 10:01 AM, Michał Prokopowicz wrote:

> @saghul https://github.com/saghul Awsome news :) Can U provide
> any estimates? (No pressure, just curious... )

We are _very_ close to 0.12. Ideally this would make it. IIRC the code
was almost  there, I'll take a stab at it this weekend if possible.

---

Saúl Ibarra Corretgé
bettercallsaghul.com

-----BEGIN PGP SIGNATURE-----
Version: GnuPG v1
Comment: Using GnuPG with Icedove - http://www.enigmail.net/

iQIcBAEBAgAGBQJTIsX3AAoJEEEOVVOum8BZLPcQAMQiRujwDiSNhPtRKDDw9rX+
e+tuOsJaMzW1rwf6COpFgxAKmuGmkm/U9LIRUBdD93v8RYYIPLk3bsp0XMTqCBa/
DGvxxZc9xYZJGp+QNXboPyjF1XWBOGflOWOmVHCmb0shABT+3KiKuB2jcv1r+9sV
iM4+/9HmliEcZROXp/mOykapTmLlVp7cH04CmBXuq2Y325TtuhBaAcxErfT1h89/
LpxPERuQFEqlsXYffPouQDf8CZEBb2ZfoK54K7Zh5vYnHa/38rjbr/WIPpEsRLvt
Yy3kvYD5IHeXqvNJ5MlsNQuh7Bd5mdt9TjB6nsJo+i3ieX393mwzPt8Hngqu07/n
L2pDHRKXI7Fpe9SNfb3PwfrKQjukymvtbhYdsTfNksGqM+HjEBFGbPv1HM3IC0RE
PGFqF0Tu0IxNBvxgnhFd9h5dRSQKD515SXmZXp5iCtcB3gFNxjRNjFIwaeyg3RcG
SOEU73qQhgC/zJ/ueWdu1C0FggDqUuu3+VWHcmQ8byv+/o0Ltpv7lG+CBgqgkXvh
HRy0SkAhyjtgvNfhDF+BvSbWnbmnKke24rZ+wNSKdAZea5c5uz/RlPc18+Q7EFbp
CLuG2b/BPnIhgLRlxOKv4IDnYBFCDABAutTLCJ1u8FY2EACGjNbMWQ1ACt0J6+di
tGuGgIt4wCLfMfVRsqSk
=YV1u
-----END PGP SIGNATURE-----

**Comment by saghul** [comment]
For those interested, I reworked the Unix part here: https://github.com/saghul/libuv/tree/set-membership-v6

I'll send a proper PR for review once I'm done with the Windows part.

**Comment by saghul** [comment]
Superseded by https://github.com/joyent/libuv/pull/1202. Lets continue the discussion there.
